File: user_interface/src/user_interface.cpp
#include "ros/ros.h"
#include "sensor_msgs/LaserScan.h"
#include "geometry_msgs/Twist.h"
#include "mode_srv/Mode.h"
#include "first_mode_srv/Firstmode.h"
#include "controller_srv/Controller.h"
#include "nav_msgs/Odometry.h"
#include <iostream>
using namespace std;

//ros::Publisher pub;
ros::ServiceClient client2;
ros::ServiceClient client1;
ros::ServiceClient client3;
int mode1=0;
int mode2=0;
int mode3=0;
int comm;
int change;
int flag = 1;
int command;

void userClk (const nav_msgs::Odometry::ConstPtr& msg)
{  
   
   mode_srv::Mode change_mode;
   first_mode_srv::Firstmode mode1_coordinates;
   controller_srv::Controller drive;
   float x;
   float y;
   
  if (flag == 1)
  {
   printf("Choose an action:\n"              
            "1 - Mode 1\n"
            "2 - Mode 2\n"
            "3 - Mode 3\n");
   
   int command;
   cin >> command;
   
   if (command == 1)
   {
   change_mode.request.type= 1;
   change_mode.request.in_mode= 1;
   client1.waitForExistence();
   client1.call(change_mode);
   mode1=1;
   mode2=0;
   mode3=0;
   }
   else if(command == 2)
   {
     change_mode.request.type= 1;
     change_mode.request.in_mode= 2;
     client1.waitForExistence();
     client1.call(change_mode);
     mode1=0;
     mode2=1;
     mode3=0;
   }
   
   else if(command == 3)
   {
     change_mode.request.type= 1;
     change_mode.request.in_mode= 3;
     client1.waitForExistence();
     client1.call(change_mode);
     mode1=0;
     mode2=0;
     mode3=1;
   }
   flag = 0;
   }
   
   if(mode1 == 1)
   {
     printf("Insert a new position \n");
     printf("New x position: \n");
     cin >> x;
     printf("New y position: \n");
     cin >> y;
     
     mode1_coordinates.request.typereq= 1;
     mode1_coordinates.request.x_in= x;
     mode1_coordinates.request.y_in= y;
     client2.waitForExistence();
     client2.call(mode1_coordinates);
     
     printf("Choose an action:\n"              
            "1 - go on with this mode\n"
            "2 - change mode 2\n");
     cin >> command;
     if (command == 2)
     {
       flag = 1;
     }       
     
   }
   
  
   
   if(mode2 == 1)
   {
     printf("\nCommands:\n"
            "1 - Go straight\n"
            "2 - Stop\n"
            "3 - Turn right\n"
            "4 - Turn left\n"
            "5 - Change mode\n"
            
            );
            
     cin >> comm;
     if (comm == 5)
     {
       flag = 1;
     }  
     else
     {
     drive.request.command= 1;
     drive.request.direction_in= comm;
     client3.waitForExistence();
     client3.call(drive);
     }
   }
   
   if(mode3 == 1)
   {
     printf("\nCommands:\n"
            "1 - Go straight\n"
            "2 - Stop\n"
            "3 - Turn right\n"
            "4 - Turn left\n"
            "5 - Change mode\n"
            );
     cin >> comm;
     
     if (comm == 5)
     {
       flag = 1;
     }  
     else
     {
     drive.request.command= 1;
     drive.request.direction_in= comm;
     client3.waitForExistence();
     client3.call(drive);
     }
   }
   
 
}



int main (int argc, char **argv)

{
   // Initialize the node, setup the NodeHandle for handling the communication with the ROS
   //system
   ros::init(argc, argv, "user_subscriber");
   ros::NodeHandle n2;
   ros::Subscriber sub = n2.subscribe("/odom", 1, userClk);
   //mode_srv::Mode change_mode;
   
   
   
   
   client1 = n2.serviceClient<mode_srv::Mode>("/mode");
   client2 = n2.serviceClient<first_mode_srv::Firstmode>("/first_mode");
   client3 = n2.serviceClient<controller_srv::Controller>("/controller");
   //pub = n2.advertise<geometry_msgs::Twist> ("/cmd_vel", 1);
   
   ros::spin();
   return 0;
}

File: user_interface/script/user_interface.py
#!/usr/bin/env python


import rospy
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Twist
from mode_srv.srv import Mode
from first_mode_srv.srv import Firstmode
from controller_srv.srv import Controller
from nav_msgs.msg import Odometry




def userClk (msg):
  
  flag = rospy.get_param("/flag")
   
  if (flag == 1):
  
           print("Choose an action:\n"              
                        "1 - Mode 1\n"
                        "2 - Mode 2\n"
                        "3 - Mode 3\n")
   
           
           command = int(input())
   
           if (command == 1):
   
                       rospy.wait_for_service("/mode")
                       client1 = rospy.ServiceProxy("/mode", Mode)
                       client1(1, 1)
                       rospy.set_param("/mode1", 1)
                       rospy.set_param("/mode2", 0)
                       rospy.set_param("/mode3", 0)
   
           elif (command == 2):
   
                       rospy.wait_for_service("/mode")
                       client1 = rospy.ServiceProxy("/mode", Mode)
                       client1(1, 2)
                       rospy.set_param("/mode1", 0)
                       rospy.set_param("/mode2", 1)
                       rospy.set_param("/mode3", 0)
   
   
           elif (command == 3):
   
                       rospy.wait_for_service("/mode")
                       client1 = rospy.ServiceProxy("/mode", Mode)
                       client1(1, 3)
                       rospy.set_param("/mode1", 0)
                       rospy.set_param("/mode2", 0)
                       rospy.set_param("/mode3", 1)
   
          
           rospy.set_param("/flag", 0)
   
   
  mode1=rospy.get_param("/mode1")
  mode2=rospy.get_param("/mode2")
  mode3=rospy.get_param("/mode3")
  if(mode1 == 1):
    
           
     
           print("Insert a new position \n")
           print("New x position: \n")
           x = float(input())
           print("New y position: \n")
           y = float(input())
     
           rospy.wait_for_service("/first_mode")
           client2 = rospy.ServiceProxy("/first_mode", Firstmode)
           client2(1, x, y)
           
           
     
           print("Choose an action:\n"              
                  "1 - go on with this mode\n"
                  "2 - change mode 2\n")
           command = int(input())
           if (command == 2):
     
                   rospy.set_param("/flag", 1)
           
     
   
   
  
   
  if(mode2 == 1):
          
           
           print("\nCommands:\n"
                  "1 - Go straight\n"
                  "2 - Stop\n"
                  "3 - Turn right\n"
                  "4 - Turn left\n"
                  "5 - Change mode\n"
            
                  )
            
           comm = int(input())
           if (comm == 5):
     
                  rospy.set_param("/flag", 1)
       
           else:
     
                  rospy.wait_for_service("/controller")
                  client3 = rospy.ServiceProxy("/controller", Controller)
                  client3(1, comm)
                  
     
   
   
  if(mode3 == 1):
   
           
           print("\nCommands:\n"
                  "1 - Go straight\n"
                  "2 - Stop\n"
                  "3 - Turn right\n"
                  "4 - Turn left\n"
                  "5 - Change mode\n"
                  )
     
           comm = int(input())
     
           if (comm == 5):
     
                  rospy.set_param("/flag", 1)
       
           else:
     
                  rospy.wait_for_service("/controller")
                  client3 = rospy.ServiceProxy("/controller", Controller)
                  client3(1, comm)
     
   
   
 




def user ():


   #Initialize the node, setup the NodeHandle for handling the communication with the ROS system
   rospy.init_node("user_subscriber")
   rospy.Subscriber("/odom", Odometry, userClk)
   
   
   
   rospy.spin()

if __name__ == '__main__':
            user()

